indigo_app/static/javascript/indigo/views/document_xml_editor.js

```diff
@@ -125,33 +165,137 @@ class AknTextEditor {
-      const xml = await this.bluebellParser.parse(
-        text,
-        this.document.get('expression_frbr_uri'),
-        fragment,
-        eidPrefix,
-      );
-
-      let newElement = $.parseXML(xml);
-
-      // if the top-level eId changed in provision mode, reload the page when saving later (just set the flag here)
-      if (this.xmlElement.parentNode.tagName === 'portionBody') {
-        // set it back to false if the eId is changed back before saving too
-        this.reloadOnSave = newElement.firstChild.firstChild.getAttribute('eId') !== Indigo.Preloads.provisionEid;
+      const elements = await this.parse();
+      // check that the response is still valid
+      if (text === this.monacoEditor.getValue()) {
+        return elements;
       }
+    } catch (err) {
+      Indigo.errorView.show(err);
+    }
 
-      if (fragmentRule === 'akomaNtoso') {
-        // entire document
-        return [newElement.documentElement];
-      } else {
-        return newElement.documentElement.children;
+    return false;
+  }
+
+  /**
+   * Accept the changes in the editor. This re-parses the text one last time and updates the DOM.
+   */
+  async acceptChanges () {
+    if (!this.editing) return;
+
+    if (!this.liveUpdates) {
+      const btn = this.root.querySelector('.btn.save');
+      btn.setAttribute('disabled', 'true');
+
+      try {
+        const elements = await this.parseSafely();
+        if (elements === false) {
+          return;
+        }
+        this.replaceElement(elements);
+      } finally {
+        btn.removeAttribute('disabled');
       }
-    } catch (e) {
-      Indigo.errorView.show(e);
     }
 
-    return null;
+    this.editing = false;
+    this.dirty = false;
+    this.xmlElement = this.xmlElementOriginal = null;
+    this.onSave();
+  }
+
+  /**
+   * Discard the changes in the editor and, if the XML element has changed, replace it with the original.
+   */
+  discardChanges () {
+    if (this.editing) {
+      // only replace the element with the original if it has changed through us
+      if (this.dirty && this.xmlElement && this.xmlElement !== this.xmlElementOriginal) {
+        this.replaceElement([this.xmlElementOriginal]);
+      }
+      this.editing = false;
+      this.dirty = false;
+      this.onDiscard();
+    }
+  }
+
+  /**
+   * Replace the current XML element with the given elements.
+   */
+  replaceElement (elements) {
+    if (this.xmlElement) {
+      if (elements && elements.length) {
+        // regular update
+        this.document.content.replaceNode(this.xmlElement, elements);
+      } else if (this.xmlElement.parentNode.tagName === 'portionBody') {
+        // we can't delete the whole provision in portion mode
+        alert($t('You cannot delete the whole provision in provision editing mode.'));
+      } else if (confirm($t('Go ahead and delete this provision from the document?'))) {
+        // remove element
+        this.document.content.replaceNode(this.xmlElement, null);
```

Comment:
would removing it rather than replacing it work better? this also can't be undone by clicking 'Cancel' after confirming deleting the provision.

Reply:
i've pushed a fix so that `getMutationChanged()` returning `removed` in a case like this. Under the hood `replaceNode(..., null)` lands up calling `.remove()`.